File: packages/PyTrilinos/doc/PyTrilinos-ACM-TOMS/test_matvecs.cpp
// Copy this file over any example that is using Triutils

#include "Epetra_ConfigDefs.h"
#ifdef HAVE_MPI
#include "mpi.h"
#include "Epetra_MpiComm.h"
#else
#include "Epetra_SerialComm.h"
#endif
#include "Epetra_Vector.h"
#include "Epetra_Time.h"
#include "Epetra_RowMatrix.h"
#include "Epetra_CrsMatrix.h"
#include "Epetra_Time.h"
#include "Epetra_LinearProblem.h"
#include "Trilinos_Util_CrsMatrixGallery.h"

using namespace Trilinos_Util;

int main(int argc, char *argv[]) 
{
#ifdef HAVE_MPI
  MPI_Init(&argc, &argv);
  Epetra_MpiComm Comm(MPI_COMM_WORLD);
#else
  Epetra_SerialComm Comm;
#endif

  CrsMatrixGallery Gallery("laplace_1d", Comm);
  Gallery.Set("problem_size",1000000 * Comm.NumProc());

  Epetra_LinearProblem* Problem = Gallery.GetLinearProblem();
  assert (Problem != 0);
  // retrive pointers to solution (lhs), right-hand side (rhs)
  // and matrix itself (A)
  Epetra_MultiVector* lhs = Problem->GetLHS();
  Epetra_MultiVector* rhs = Problem->GetRHS();
  Epetra_RowMatrix* A = Problem->GetMatrix();
    
  Epetra_Time Time(Comm);

  for (int i = 0 ; i < 100 ; ++i)
    A->Multiply(false, *lhs, *rhs);

  cout << Time.ElapsedTime() << endl;

  return(EXIT_SUCCESS);
} // end of main()

File: packages/PyTrilinos/doc/PyTrilinos-ACM-TOMS/test_matvecs.py
#! /usr/bin/env python
from PyTrilinos import Epetra, Triutils

Epetra.Init()
Comm = Epetra.PyComm()
n = 1000000 * Comm.NumProc()
Gallery = Triutils.CrsMatrixGallery("laplace_1d", Comm)
Gallery.Set("problem_size", n)
Matrix = Gallery.GetMatrix()
LHS = Gallery.GetStartingSolution()
RHS = Gallery.GetRHS()
Time = Epetra.Time(Comm)

for i in xrange(100):
  Matrix.Multiply(False, LHS, RHS)

print Time.ElapsedTime()
Epetra.Finalize()
